### PR TITLE
Alias and Index Rollover changes

### PR DESCRIPTION
**Pull request by ashish-khatkar**

### CODE REVIEW

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/IndexRollOverManagerConfig.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+import java.util.Map;
+
+/**
+ * Created by ashish.khatkar on 12/12/17.
+ */
+public class IndexRollOverManagerConfig {
+    @Min(3600)
+    private int interval;
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/IndexRollOverManagerConfig.java]: Put timeUnit in variable names. intervalInSeconds for eg.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/IndexRollOverManagerConfig.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Min(1)
+    private int initialDelay;
+
+    @NotNull
+    private boolean active;
+
+    private Map<String, Object> conditions;
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/IndexRollOverManagerConfig.java]: Can this be modelled better ?

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/IndexAliasManager.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+import java.util.List;
+
+/**
+ * Created by ashish.khatkar on 12/12/17.
+ */
+public interface IndexAliasManager extends Managed {
+    void save(String aliasName) throws FoxtrotException;
+    boolean exists(String aliasName) throws FoxtrotException;
+    List<String> get() throws FoxtrotException;
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/IndexAliasManager.java]: expand the names a bit.
getAliases(), aliasExists() etc..

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/IndexRollOverTask.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        this.conditions = conditions;
+    }
+
+    @Override
+    public void run() {
+        logger.info("Starting index rollover for job");
+        try {
+            queryStore.indexRollOver(conditions);
+        } catch (FoxtrotException ex) {
+            logger.error("Index rollover job failed", ex);
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/IndexRollOverTask.java]: Need:
1) TimeStats on rollOver
2) Metrics on init, success and failure of the periodic tasks

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/CountAction.java [hunk 5]
```diff
@@ -95,7 +95,7 @@ public ActionResponse execute(CountRequest parameter) throws FoxtrotException {
             SearchRequestBuilder query;
             try {
                 query = getConnection().getClient()
-                        .prepareSearch(ElasticsearchUtils.getIndices(parameter.getTable(), parameter))
+                        .prepareSearch(ElasticsearchUtils.getIndicesForSearch(parameter.getTable(), parameter))
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/actions/CountAction.java:98]: Deprecate / Remove getIndices() if not needed.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.List;
+
+/**
+ * Created by ashish.khatkar on 12/12/17.
+ */
+public class DistributedIndexAliasManager implements IndexAliasManager {
+    private static final Logger logger = LoggerFactory.getLogger(DistributedIndexAliasManager.class);
+    private static final String ALIAS_MAP = "indexaliasmap";
+    private final HazelcastConnection hazelcastConnection;
+    private IMap<String, String> indexAliasStore;
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java]: Replace with List if Map isn't needed.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Created by ashish.khatkar on 12/12/17.
+ */
+public class DistributedIndexAliasManager implements IndexAliasManager {
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java:29]: Review pending, post IMap changes.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.EAGER);
+        MapConfig mapConfig = new MapConfig(ALIAS_MAP);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        mapConfig.setReadBackupData(true);
+        hazelcastConnection.getHazelcastConfig().addMapConfig(mapConfig);
+        DistributedCache.setupConfig(hazelcastConnection);
+    }
+
+    @Override
+    public void save(String aliasName) throws FoxtrotException {
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java]: Put metrics on failures (success if necessary).

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/IndexAliasMapStore.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+
+
+/**
+ * Created by ashish.khatkar on 12/12/17.
+ */
+public class IndexAliasMapStore implements MapStore<String, String> {
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/IndexAliasMapStore.java:21]: Review Pending post iMap removal.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        logger.info("Saving alias {}", aliasName);
+        indexAliasStore.put(aliasName, aliasName);
```
Comment (ashish-khatkar) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java]: Have a hazelcast lock here. put and flush should be done in a single transaction.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/IndexRollOverManager.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    final Timer timer;
+
+    public IndexRollOverManager(IndexRollOverManagerConfig config, QueryStore queryStore) {
+        this.config = config;
+        this.querystore = queryStore;
+        this.timer = new Timer(true);
+    }
+
+    @Override
+    public void start() throws Exception {
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/IndexRollOverManager.java]: Remove Exception from declaration if not needed.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                                        ElasticsearchConnection elasticsearchConnection) {
+        this.hazelcastConnection = hazelcastConnection;
+        this.elasticsearchConnection = elasticsearchConnection;
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setFactoryImplementation(IndexAliasMapStore.factory(elasticsearchConnection));
+        mapStoreConfig.setEnabled(true);
+        mapStoreConfig.setInitialLoadMode(MapStoreConfig.InitialLoadMode.EAGER);
+        MapConfig mapConfig = new MapConfig(ALIAS_MAP);
+        mapConfig.setMapStoreConfig(mapStoreConfig);
+        mapConfig.setReadBackupData(true);
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/DistributedIndexAliasManager.java:47]: 1.Check this once.
2. Replication of In-Memory Map. Check config for it.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/ElasticsearchUtils.java [hunk 13]
```diff
@@ -247,4 +247,76 @@ public static String getTableNameFromIndex(String currentIndex) {
...
+    /**
+     * Removes -alias from the alias name to get starting for new index
+     * For ex : if alias is foxtrot-appName-table-12-12-2017-alias, corresponding initial index will be foxtrot-appName-table-12-12-2017-000001
+     * @param alias : Name of the alias
+     * @return string for the first index to which alias should point
+     */
+    public static String getIndexFromAlias(String alias) {
+        if (alias.contains(INDEX_ALIAS)) {
+            int position = alias.lastIndexOf(String.format("-%s", INDEX_ALIAS));
+            return String.format("%s-%s", alias.substring(0, position), INITIAL_INDEX);
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/ElasticsearchUtils.java:260]: move it to public static final

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/IndexAliasMapStore.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    public static Factory factory(ElasticsearchConnection elasticsearchConnection) {
+        return new Factory(elasticsearchConnection);
+    }
+
+    private final ElasticsearchConnection elasticsearchConnection;
+    private final ObjectMapper objectMapper;
+
+    public IndexAliasMapStore(ElasticsearchConnection elasticsearchConnection) {
+        this.elasticsearchConnection = elasticsearchConnection;
+        this.objectMapper = new ObjectMapper();
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/IndexAliasMapStore.java]: Remove OM. Not being used.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/IndexAliasMapStore.java [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+    }
+
+    @Override
+    public void deleteAll(Collection<String> keys) {
+
+    }
+
+    @Override
+    public String load(String key) {
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/impl/IndexAliasMapStore.java]: Leave a note/documentation on why load is returning null but loadAll is not.

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/AliasConditions.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
+package com.flipkart.foxtrot.core.common;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.validation.constraints.Min;
+
```
Comment (shivam-rai) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/AliasConditions.java:6]: Add no. of shards to config.

**Pull request merged**